lib/spack/spack/test/concretize.py

```diff
@@ -1836,3 +1836,33 @@ def test_installed_externals_are_reused(
         with spack.config.override("concretizer:reuse", True):
             external3 = Spec('changing@1.0').concretized()
         assert external3.dag_hash() == external1.dag_hash()
+
+    @pytest.mark.regression('31484')
+    def test_installed_builds_not_reused_buildable_false(
+            self, mutable_database, repo_with_changing_recipe
+    ):
```

Comment:
@becker33 I have been getting back at this specific test, and I think this is behavior we don't want to add as it is coded here. The only things we  currently say for `buildable: false` in our documentation are:

> The addition of the buildable flag tells Spack that it should never build its own version of OpenMPI, and it will instead always rely on a pre-built OpenMPI.
>
> The buildable does not need to be paired with external packages. It could also be used alone to forbid packages that may be buggy or otherwise undesirable.
> [ ... ]

and they clearly stem from a world where `reuse` was not a considered at all. It seems thus a stretch to me to require that `buildable: false` prevents the reuse of software that is either installed locally or available from a buildcache.

That said, I concur, based on our discussion on Slack, that there's a gap in expressiveness that we need to fill. For instance, as you pointed out:
```yaml
packages:
  mpi:
    buildable: false
```
likely means that the user wants to install only from one of the externals that might be listed in the same file. But what if it is:
```yaml
packages:
  trilinos:
    buildable: false
```
Couldn't that be a way to say "Reuse a binary version of trilinos and never build that from sources"? I wouldn't find that request "crazy", if a binary cache is available. 

I think the point is: `buildable: false` was added before `reuse` and we didn't figure out how the two should interact together. That is possibly what needs to get fixed (allow the user to express in which of the two cases they are), not excluding specs from reuse based on `buildable: false`.

**TL;DR** I agree that here we miss ways to express a few requests, but would it be possible to get the bugfix in as is, and work on a more extensive feature in a following PR? lf not I'm tempted to move this out of the v0.18.1 project, since it likely means it won't be solved soon.

Reply:
> I still don't see how someone using this is supposed to use system mpi, and I don't think that's a use case we can afford to break.

Fair, I already moved this to the next bugfix release. Do you agree though that the test, as is, is not something we want to force implicitly?